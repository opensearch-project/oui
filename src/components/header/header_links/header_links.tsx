/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  HTMLAttributes,
  FunctionComponent,
  useState,
  useEffect,
  MouseEventHandler,
} from 'react';
import classNames from 'classnames';

import { CommonProps, keysOf } from '../../common';
import { OuiIcon, IconType } from '../../icon';
import { OuiPopover, OuiPopoverProps } from '../../popover';
import { OuiI18n } from '../../i18n';
import {
  OuiHeaderSectionItemButton,
  OuiHeaderSectionItemButtonProps,
} from '../header_section';
import { OuiBreakpointSize } from '../../../services/breakpoint';
import { OuiHideFor, OuiShowFor } from '../../responsive';

type OuiHeaderLinksGutterSize = 'xs' | 's' | 'm' | 'l';
type OuiHeaderLinksPopoverButtonProps = Partial<
  OuiHeaderSectionItemButtonProps
> & {
  iconType?: IconType;
};

export type OuiHeaderLinksProps = CommonProps &
  HTMLAttributes<HTMLElement> & {
    /**
     * Spacing between direct children
     */
    gutterSize?: OuiHeaderLinksGutterSize;
    /**
     * A list of named breakpoints at which to show the popover version
     */
    popoverBreakpoints?: OuiBreakpointSize[] | 'all' | 'none';
    /**
     * Extend the functionality of the OuiPopover.button which is a OuiHeaderSectionItemButton.
     * With the addition of `iconType` to change the display icon which defaults to `apps`
     */
    popoverButtonProps?: OuiHeaderLinksPopoverButtonProps;
    /**
     * Extend the functionality of the OuiPopover
     */
    popoverProps?: Omit<OuiPopoverProps, 'button' | 'closePopover'>;
  };

const gutterSizeToClassNameMap: {
  [gutterSize in OuiHeaderLinksGutterSize]: string;
} = {
  xs: '--gutterXS',
  s: '--gutterS',
  m: '--gutterM',
  l: '--gutterL',
};
export const GUTTER_SIZES = keysOf(gutterSizeToClassNameMap);

export const OuiHeaderLinks: FunctionComponent<OuiHeaderLinksProps> = ({
  children,
  className,
  gutterSize = 's',
  popoverBreakpoints = ['xs', 's'],
  popoverButtonProps,
  popoverProps,
  ...rest
}) => {
  const { onClick, iconType = 'apps', ...popoverButtonRest } =
    popoverButtonProps ?? {};

  const [mobileMenuIsOpen, setMobileMenuIsOpen] = useState(false);

  const onMenuButtonClick: MouseEventHandler<
    HTMLButtonElement & HTMLAnchorElement
  > = (e) => {
    onClick?.(e);
    setMobileMenuIsOpen(!mobileMenuIsOpen);
  };

  const closeMenu = () => {
    setMobileMenuIsOpen(false);
  };

  useEffect(() => {
    window.addEventListener('resize', closeMenu);
    return () => {
      window.removeEventListener('resize', closeMenu);
    };
  });

  const classes = classNames('ouiHeaderLinks', className);

  const button = (
    <OuiI18n token="ouiHeaderLinks.openNavigationMenu" default="Open menu">
      {(openNavigationMenu: string) => (
        <OuiHeaderSectionItemButton
          aria-label={openNavigationMenu}
          onClick={onMenuButtonClick}
          {...popoverButtonRest}>
          <OuiIcon type={iconType} size="m" />
        </OuiHeaderSectionItemButton>
      )}
    </OuiI18n>
  );

  return (
    <OuiI18n token="ouiHeaderLinks.appNavigation" default="App menu">
      {(appNavigation: string) => (
        <nav className={classes} aria-label={appNavigation} {...rest}>
          <OuiHideFor sizes={popoverBreakpoints}>
            <div
              className={classNames('ouiHeaderLinks__list', [
                `ouiHeaderLinks__list${gutterSizeToClassNameMap[gutterSize]}`,
              ])}>
              {children}
            </div>
          </OuiHideFor>

          <OuiShowFor sizes={popoverBreakpoints}>
            <OuiPopover
              button={button}
              isOpen={mobileMenuIsOpen}
              anchorPosition="downRight"
              closePopover={closeMenu}
              panelPaddingSize="none"
              {...popoverProps}>
              <div
                className={classNames('ouiHeaderLinks__mobileList', [
                  `ouiHeaderLinks__mobileList${gutterSizeToClassNameMap[gutterSize]}`,
                ])}>
                {children}
              </div>
            </OuiPopover>
          </OuiShowFor>
        </nav>
      )}
    </OuiI18n>
  );
};
