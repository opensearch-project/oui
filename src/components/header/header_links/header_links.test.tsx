/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../../test';

import { OuiHeaderLinks, GUTTER_SIZES } from './header_links';

describe('OuiHeaderLinks', () => {
  test('is rendered', () => {
    const component = render(<OuiHeaderLinks {...requiredProps} />);

    expect(component).toMatchSnapshot();
  });

  describe('gutterSize', () => {
    GUTTER_SIZES.forEach((gutterSize) => {
      test(`${gutterSize} is rendered`, () => {
        const component = render(<OuiHeaderLinks gutterSize={gutterSize} />);

        expect(component).toMatchSnapshot();
      });
    });
  });

  describe('popover props', () => {
    test('is rendered', () => {
      const component = render(
        <OuiHeaderLinks
          popoverBreakpoints={['xs', 's', 'm', 'l', 'xl', 'xxl', 'xxxl']}
          popoverButtonProps={{
            iconType: 'bolt',
            className: 'customButtonClass',
          }}
          popoverProps={{ anchorClassName: 'customAnchorClass' }}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('is never rendered with "none"', () => {
      const component = render(<OuiHeaderLinks popoverBreakpoints={'none'} />);

      expect(component).toMatchSnapshot();
    });
  });
});
