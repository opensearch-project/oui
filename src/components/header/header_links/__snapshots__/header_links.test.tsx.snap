// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiHeaderLinks gutterSize l is rendered 1`] = `
<nav
  aria-label="App menu"
  class="ouiHeaderLinks"
>
  <div
    class="ouiHeaderLinks__list ouiHeaderLinks__list--gutterL"
  />
</nav>
`;

exports[`OuiHeaderLinks gutterSize m is rendered 1`] = `
<nav
  aria-label="App menu"
  class="ouiHeaderLinks"
>
  <div
    class="ouiHeaderLinks__list ouiHeaderLinks__list--gutterM"
  />
</nav>
`;

exports[`OuiHeaderLinks gutterSize s is rendered 1`] = `
<nav
  aria-label="App menu"
  class="ouiHeaderLinks"
>
  <div
    class="ouiHeaderLinks__list ouiHeaderLinks__list--gutterS"
  />
</nav>
`;

exports[`OuiHeaderLinks gutterSize xs is rendered 1`] = `
<nav
  aria-label="App menu"
  class="ouiHeaderLinks"
>
  <div
    class="ouiHeaderLinks__list ouiHeaderLinks__list--gutterXS"
  />
</nav>
`;

exports[`OuiHeaderLinks is rendered 1`] = `
<nav
  aria-label="aria-label"
  class="ouiHeaderLinks testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    class="ouiHeaderLinks__list ouiHeaderLinks__list--gutterS"
  />
</nav>
`;

exports[`OuiHeaderLinks popover props is never rendered with "none" 1`] = `
<nav
  aria-label="App menu"
  class="ouiHeaderLinks"
>
  <div
    class="ouiHeaderLinks__list ouiHeaderLinks__list--gutterS"
  />
</nav>
`;

exports[`OuiHeaderLinks popover props is rendered 1`] = `
<nav
  aria-label="App menu"
  class="ouiHeaderLinks"
>
  <div
    class="ouiPopover ouiPopover--anchorDownRight"
  >
    <div
      class="ouiPopover__anchor customAnchorClass"
    >
      <button
        aria-label="Open menu"
        class="ouiButtonEmpty ouiButtonEmpty--text ouiHeaderSectionItemButton customButtonClass"
        type="button"
      >
        <span
          class="ouiButtonContent ouiButtonEmpty__content"
        >
          <span
            class="ouiButtonEmpty__text"
          >
            <span
              class="ouiHeaderSectionItemButton__content"
            >
              <span
                data-ouiicon-type="bolt"
              />
            </span>
          </span>
        </span>
      </button>
    </div>
  </div>
</nav>
`;
